rfcs/text/0009_screenshot_mode_service.md

```diff
@@ -0,0 +1,262 @@
+background to inject the data into the various running stages of applications.
+This is in line with the pattern on the client-side where all query parameters
+are delegated to the individual applications, letting them do as they wish with
+it. For this RFC, it's an important topic because plugins have to "wire-up"
+this service in their router and UI app, but it will prevent "magical" behavior
+happening.
+
+## Server side
+On the server side, applications will have to interact with the screenshot
+plugin to support custom endpoints include the screenshot data as part of the
+route handling context. Example:
+```
+const screenshotRouter = plugins.screenshots.wrapRouter(router);
+screenshotRouter.get(
+  { path: '/my-endpoint' },
+  async (context, req, res) => {
+    const isScreenshotMode = context.screenshots.isScreenshotMode();
+  }
+);
+```
+
+The `screenshotRouter` handlers could either replace the existing routes of the
+application, or the application can feature custom endpoints for screenshot
+export that use this plugin.
+
+## Front-end
+On the front-end, the plugin service could be built as a simple utility that takes a
+[History](https://developer.mozilla.org/en-US/docs/Web/API/History)
+instance and returns an Observable that emits the screenshot data by listening
+for location updates and computing the data based on the query parameters.
+
+```
+const screenshot$ = plugins.screenshots.readFrom(history);
+```
+
+## Types
+The `request.screenshotMode` object is an instance of a class that has
+functions to abstract it's internal raw data:
+
+```
+class ScreenshotModeServiceApi {
+  constructor(rawData) {
+  }
+  public isScreenshotMode (): boolean {
+  }
+  public getScreenshotDimensions (): ScreenshotLayoutDimensions {
+  }
+  public getTimezone (): Timezone {
+  }
+}
+```
+
+Getting the screenshot dimensions would be needed only for very specialized use
+cases such as the Dashboard application's "print layout" mode.
+
+# Drawbacks
+
+Why should we *not* do this? Please consider:
+
+- Plugins have to "wire-up" this service in their router and UI app
+- Low-level of discoverability since this logic is hidden in a plugin rather than exposed on the CoreSetup API.
+- Hard for application teams to create an environment to test against to check the screenshot mode rendering of their work.
+
+As a solution to the drawbacks of extra maintenance and tests needed, the
+Reporting Services team could provide libraries to use in automated functional
+tests that help verify the integration. There could also be a development tool
+in the form of a test Kibana plugin or bookmarklet to help developers check their
+work for screenshot mode.
+
+# Alternatives
+
+- Print media query CSS
+  Many things can be improved on capturing screenshots from the page using
+  `@print` CSS media query selecting. Also, CSS makes things hidden on the
+  page, but they are still loading in the DOM and take browser memory
```

Comment:
Is it a really problem? AFAIK you can command a browser not to load styles using [media attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/link):
```html
<link href="print.css" rel="stylesheet" media="print">
```

Reply:
The reason we will use print media query CSS as a supplement is because we can't address page loading performance using just this solution. We need faster page loading performance than we have today to deliver the reports in a timely manner.

Reply:
How can I better understand the performance issues we face here with Reporting where media queries aren't sufficient. As I understand it, any performance issues we have in Reporting are also issues every-day users experience (bundle size, etc). I am concerned that the RFC suggestion doesn't address the underlying issue, but would like to understand further.

Reply:
@tylersmalley 

My understanding is also that a lot of the problems happening in Reporting are issues every-day users face. I doubt that we can do anything about bundle sizes, even if this service existed. But I think we can find ways to decrease memory usage for non-interactive experiences. It would be wonderful if this helps reports run faster, but we're also still very concerned about stability and we think that streamlining where we can will help more reports complete without Chromium running out of memory. 

Having this service in place would give us leverage to streamline more things. The background polling requests seem to be the primary negative factor on Reporting performance, especially higher memory consumption. We're working with the idea that the plugins that do background polling can be adjusted for screenshot capture, where only the initial data should be needed.

We can't quantify how much of a hit that it's taking on performance because we don't have the tools to isolate that. The Screenshot Mode Service would be the tool that would let us tweak the screenshot capture experience. Being able to flip some switches would help us understand more about what is impacting performance, while [we measure things happening with APM](https://github.com/elastic/kibana/pull/5996).